src/merkle/partial_mt/mod.rs

```diff
@@ -0,0 +1,365 @@
+    pub fn is_leaf(&self, index: NodeIndex) -> bool {
+        self.leaves.contains(&index)
+    }
+
+    pub fn get_leaf_depth(&self, index: u64) -> Result<u8, MerkleError> {
+        let mut node_index = NodeIndex::new(self.max_depth(), index)?;
+        for _ in 0..node_index.depth() {
+            if self.leaves.contains(&node_index) {
+                return Ok(node_index.depth());
+            }
+            node_index.move_up()
+        }
+        // we don't have an error for this case, maybe it makes sense to create a new error, something like
+        // NoLeafForIndex("There is no leaf for provided index"). But it will be used almost never.
+        Err(MerkleError::NodeNotInSet(node_index))
+    }
+
+    /// Returns a value of the leaf at the specified NodeIndex.
+    ///
+    /// # Errors
+    /// Returns an error if the NodeIndex is not contained in the leaves set.
+    pub fn get_leaf(&self, index: NodeIndex) -> Result<Word, MerkleError> {
+        if !self.leaves.contains(&index) {
+            // This error not really suitable in this situation, should I create a new error?
+            Err(MerkleError::InvalidIndex {
+                depth: index.depth(),
+                value: index.value(),
+            })
+        } else {
+            self.nodes
+                .get(&index)
+                .ok_or(MerkleError::NodeNotInSet(index))
+                .map(|hash| **hash)
+        }
+    }
+
+    /// Returns a map of paths from every leaf to the root.
+    pub fn paths(&self) -> Result<BTreeMap<&NodeIndex, MerklePath>, MerkleError> {
+        let mut paths = BTreeMap::new();
+        for leaf_index in self.leaves.iter() {
+            let index = *leaf_index;
+            paths.insert(leaf_index, self.get_path(index)?);
+        }
+        Ok(paths)
+    }
+
+    /// Returns a Merkle path from the node at the specified index to the root.
+    ///
+    /// The node itself is not included in the path.
+    ///
+    /// # Errors
+    /// Returns an error if:
+    /// - the specified index has depth set to 0 or the depth is greater than the depth of this
+    /// Merkle tree.
+    /// - the specified index is not contained in the nodes map.
+    pub fn get_path(&self, mut index: NodeIndex) -> Result<MerklePath, MerkleError> {
+        if index.is_root() {
+            return Err(MerkleError::DepthTooSmall(index.depth()));
+        } else if index.depth() > self.max_depth() {
+            return Err(MerkleError::DepthTooBig(index.depth() as u64));
+        }
+
+        if !self.nodes.contains_key(&index) {
+            return Err(MerkleError::NodeNotInSet(index));
+        }
+
+        let mut path = Vec::new();
+        for _ in 0..index.depth() {
+            let sibling_index = Self::get_sibling_index(&index)?;
+            index.move_up();
+            let sibling_hash =
+                self.nodes.get(&sibling_index).cloned().unwrap_or(Self::EMPTY_DIGEST);
```

Comment:
Why are we using `EMPTY_DIGEST` as default here? If a node is in the tree, then its path should aways be in the tree as well, no? So, the situation when sibling is not in the tree should never arise. If so, we should just use `.expect()` here.

Also, minor nit: I'd name the variable just `sibling` (rather than `sibling_hash`).